[Libp2p] Add libp2p module directories and helpers (part 1)

bryanchriswhite opened the pull request:
<!-- REMOVE this comment block after following the instructions
 1. Make the title of the PR is descriptive and follows this format: `[<Module>] <DESCRIPTION>`
 2. Update the _Assigness_, _Labels_, _Projects_, _Milestone_ before submitting the PR for review.
 3. Add label(s) for the purpose (e.g. `persistence`) and, if applicable, priority (e.g. `low`) labels as well.
-->

## Description

This is the first of a series of PRs split out from #500. Here we set up the new libp2p module directory structure and add crypto and identity / network helpers.

## Issue

#347 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- prepare pocket repo new libp2p module
- add pocket / libp2p identity helpers
- add url <--> multiaddr conversion helpers for use with libp2p (see: https://github.com/multiformats/go-multiaddr)
- add pokt --> libp2p crypto helpers

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

<!-- REMOVE this comment block after following the instructions
 If you added additional tests or infrastructure, describe it here.
 Bonus points for images and videos or gifs.
-->

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

bryanchriswhite commented:
## Changelog validation failure

I think this is a false negative caused by the changelog in question having no previous version (as it's new). You can apply this diff locally to prove this:

```
diff --git i/libp2p/docs/CHANGELOG.md w/libp2p/docs/CHANGELOG.md
index 4de971d9..e533805b 100644
--- i/libp2p/docs/CHANGELOG.md
+++ w/libp2p/docs/CHANGELOG.md
@@ -5,7 +5,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
-## [0.0.0.0] - 2023-02-22
+## [0.0.0.1] - 2023-02-22
 
 - prepare pocket repo new libp2p module
 - add pocket / libp2p identity helpers
@@ -14,3 +14,5 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - add `Multiaddr` field to `typesP2P.NetworkPeer`
 
 <!-- GITHUB_WIKI: changelog/libp2p -->
+
+## [0.0.0.0] - 2023-02-22
```

## Test failure

I am unable to reproduce the behavior seen in CI locally (i.e. exit code 1). I am also, apparently, unable to locate any error in the output provided by CI; I tried searching for "fail", "error", and manual, visual inspection. I double-checked that indeed what I'm running locally and what CI are doing are ultimately the same. [CI is calling `go test ./...`](https://github.com/pokt-network/pocket/blob/main/.github/workflows/main.yml#L58) (plus some other args), and [`make develop_test`](https://github.com/pokt-network/pocket/blob/main/Makefile#L122) calls [`make test_all`](https://github.com/pokt-network/pocket/blob/main/Makefile#L317) which also runs `go test ./...` (again, plus some other args) so I'm confused at the moment.

bryanchriswhite commented:
I discovered this after looking through the raw log output:

```
2023-02-22T18:59:58.9141066Z {"Time":"2023-02-22T18:59:58.913826591Z","Action":"output","Package":"github.com/pokt-network/pocket/persistence/test","Test":"TestStateHash_DeterministicStateWhenUpdatingAppStake","Output":"--- FAIL: TestStateHash_DeterministicStateWhenUpdatingAppStake (0.77s)\n"}
2023-02-22T18:59:58.9145608Z {"Time":"2023-02-22T18:59:58.914367396Z","Action":"fail","Package":"github.com/pokt-network/pocket/persistence/test","Test":"TestStateHash_DeterministicStateWhenUpdatingAppStake","Elapsed":0.77}
```

bryanchriswhite commented:
## Test failing in CI but not locally

I've extracted the relevant lines from the raw logs, it's not clear to me why there is no additional output from whatever assertion supposedly failed:
```
2023-02-23T09:45:38.4162036Z {"Time":"2023-02-23T09:45:38.413936184Z","Action":"output","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error","Output":"--- FAIL: TestPeerMultiAddrFromServiceURL_error (0.03s)\n"}
2023-02-23T09:45:38.4163263Z {"Time":"2023-02-23T09:45:38.413942784Z","Action":"output","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/fqdn","Output":"    --- FAIL: TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/fqdn (0.01s)\n"}
2023-02-23T09:45:38.4164338Z {"Time":"2023-02-23T09:45:38.413946184Z","Action":"fail","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/fqdn","Elapsed":0.01}
2023-02-23T09:45:38.4165564Z {"Time":"2023-02-23T09:45:38.413968185Z","Action":"output","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip4","Output":"    --- FAIL: TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip4 (0.01s)\n"}
2023-02-23T09:45:38.4166675Z {"Time":"2023-02-23T09:45:38.413971985Z","Action":"fail","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip4","Elapsed":0.01}
2023-02-23T09:45:38.4167818Z {"Time":"2023-02-23T09:45:38.413975785Z","Action":"output","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip6","Output":"    --- FAIL: TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip6 (0.01s)\n"}
```

In contrast, here is an example from the raw logs (of an unrelated test failure), where you see the stack trace in "output" lines as you would expect:
```
network/pocket/consensus/e2e_tests","Test":"TestPacemakerCatchupSameStepDifferentRounds","Output":"--- FAIL: TestPacemakerCatchupSameStepDifferentRounds (0.02s)\n"}
2023-02-23T09:04:21.7728306Z {"Time":"2023-02-23T09:04:21.772196843Z","Action":"output","Package":"github.com/pokt-network/pocket/consensus/e2e_tests","Test":"TestPacemakerCatchupSameStepDifferentRounds","Output":"panic: reflect: Call using *types.NodeId as type uint64 [recovered]\n"}
2023-02-23T09:04:21.7729572Z {"Time":"2023-02-23T09:04:21.772213743Z","Action":"output","Package":"github.com/pokt-network/pocket/consensus/e2e_tests","Test":"TestPacemakerCatchupSameStepDifferentRounds","Output":"\tpanic: reflect: Call using *types.NodeId as type uint64\n"}
2023-02-23T09:04:21.7730474Z {"Time":"2023-02-23T09:04:21.772219643Z","Action":"output","Package":"github.com/pokt-network/pocket/consensus/e2e_tests","Test":"TestPacemakerCatchupSameStepDifferentRounds","Output":"\n"}
2023-02-23T09:04:21.7731334Z {"Time":"2023-02-23T09:04:21.772223543Z","Action":"output","Package":"github.com/pokt-network/pocket/consensus/e2e_tests","Test":"TestPacemakerCatchupSameStepDifferentRounds","Output":"goroutine 136 [running]:\n"}
2023-02-23T09:04:21.7732285Z {"Time":"2023-02-23T09:04:21.772227343Z","Action":"output","Package":"github.com/pokt-network/pocket/consensus/e2e_tests","Test":"TestPacemakerCatchupSameStepDifferentRounds","Output":"testing.tRunner.func1.2({0xb0a460, 0xc000377e00})\n"}
2023-02-23T09:04:21.7733681Z {"Time":"2023-02-23T09:04:21.772231043Z","Action":"output","Package":"github.com/pokt-network/pocket/consensus/e2e_tests","Test":"TestPacemakerCatchupSameStepDifferentRounds","Output":"\t/opt/hostedtoolcache/go/1.19.5/x64/src/testing/testing.go:1396 +0x24e\n"}
2023-02-23T09:04:21.7734620Z {"Time":"2023-02-23T09:04:21.772235143Z","Action":"output","Package":"github.com/pokt-network/pocket/consensus/e2e_tests","Test":"TestPacemakerCatchupSameStepDifferentRounds","Output":"testing.tRunner.func1()\n"}
2023-02-23T09:04:21.7735610Z {"Time":"2023-02-23T09:04:21.772238343Z","Action":"output","Package":"github.com/pokt-network/pocket/consensus/e2e_tests","Test":"TestPacemakerCatchupSameStepDifferentRounds","Output":"\t/opt/hostedtoolcache/go/1.19.5/x64/src/testing/testing.go:1399 +0x39f\n"}
```

Olshansk commented on shared/crypto/libp2p.go:
```diff
@@ -0,0 +1,23 @@
+
+	privateKey, err := crypto.PrivKeyUnmarshallers[crypto.Ed25519](privateKeyBz)
+	if err != nil {
+		return nil, ErrCreatePublicKey(err)
```
Maybe `ErrCreatePrivateKey` instead?

bryanchriswhite replied:
Nice catch! :raised_hands: I think that's an autocomplete typo.

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -0,0 +1,107 @@
+package network
```
Nice naming scehemes for the files!

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -0,0 +1,107 @@
+		)
+	}
+
+	var (
```
Thoughts on putting this at the top of the file (outside the function)?

It'll help capture the TECHDEBT and current state of things immediately.

bryanchriswhite replied:
For sure, there's no reason for this to be reallocated each call - I just missed this. :sweat_smile:  Nice catch! :raised_hands:

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -0,0 +1,107 @@
+
+	// TECHDEBT: there's probably a more conventional way to do this.
+	// Check if we're dealing with IPv4 or IPv6
+	if strings.Count(peerUrl.Hostname(), ":") > 0 {
```
Seems like we can just use the .net package and convert to Ip4. If it's nil, it's ip6 https://pkg.go.dev/net#IP.To4

bryanchriswhite replied:
Nice one!

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -0,0 +1,107 @@
+	 * TECHDEBT: is there a way for us to effectively prefer IPv6 responses when resolving DNS?
+	 * TECHDEBT: resolving DNS this way has limitations:
+	 * > The address parameter can use a host name, but this is not recommended,
+	 * > because it will return at most one of the host name's IP addresses.
```
Which one specifically? Just a random one?

bryanchriswhite replied:
I would have to read a bit through stdlib to be able to answer that, the lines prefixed with `> `  are copy/pasted from the doc page on the line below.

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -0,0 +1,107 @@
+
+// ServiceUrlFromLibp2pMultiaddr converts a multiaddr into a URL string.
+func ServiceUrlFromLibp2pMultiaddr(addr multiaddr.Multiaddr) (string, error) {
+	protos := addr.Protocols()
```
NIT: Can you rename `proto` in this function to `protocol` or `prtcl`? I was reading it as `protobuf` which we use throughout the codebase.

Olshansk commented on libp2p/network/url_conversion_test.go:
```diff
@@ -0,0 +1,159 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestPeerMultiAddrFromServiceURL_success(t *testing.T) {
```
```suggestion
func TestPeerMultiAddrFromServiceURL_Success(t *testing.T) {
```

This is super helpful, ty!

Olshansk commented on libp2p/network/url_conversion_test.go:
```diff
@@ -0,0 +1,159 @@
+	}
+}
+
+const (
```
Where are we using these?

bryanchriswhite replied:
Just in this test. I'll unexport them. :wink:

Olshansk commented on libp2p/network/url_conversion_test.go:
```diff
@@ -0,0 +1,159 @@
+	IP6  = "ip6"
+)
+
+func TestPeerMultiAddrFromServiceURL_error(t *testing.T) {
```
```suggestion
func TestPeerMultiAddrFromServiceURL_Error(t *testing.T) {
```

Ditto below

Olshansk commented on libp2p/network/url_conversion_test.go:
```diff
@@ -0,0 +1,159 @@
+			require.NoError(t, err)
+
+			_, err = ServiceUrlFromLibp2pMultiaddr(addr)
+			// DISCUSS: asserting specific errors instead.
```
I think this is reasonable. Did you want to discuss if we should do it everywhere?

bryanchriswhite replied:
Yes, I think it would be good to discuss it. I'm still developing my own opinion but at the moment I would prefer not to necessarily encourage emulating this. 

I think before we put too much effort into action, it would be prudent to nail down some reasonable requirements / acceptance criteria for errors / error handling conventions. I've started issue #519 for this purpose. In absence of such requirements, the intention behind wrapping errors like I've been doing here is mainly to aid in debugging.

bryanchriswhite replied:
Dave Cheney (in a [GopherCon 2016 presentation](https://youtu.be/lsBF58Q-DnY?t=701)) talks about preferring to assert for errors by behavior over type which mitigates problems resulting from other solutions. So perhaps, asserting for specific errors isn't quite the right direction.

Olshansk replied:
> I've started issue #519 for this purpose. In absence of such requirements, the intention behind wrapping errors like I've been doing here is mainly to aid in debugging.

Noted. Seems like there's nothing actionable to discuss in this commit, but will keep it in mind.

Olshansk reviewed (changes requested):
A few minor comments, but for the most part super clean and well compartmentalized as a separate PR. Appreciate making it easy for the reviewer

bryanchriswhite commented:
Despite merging @deblasis's change to print more output in CI, a subset of test cases continue to fail with no output in CI:

```
2023-02-23T16:22:46.7367092Z {"Time":"2023-02-23T16:22:46.735690851Z","Action":"output","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error","Output":"--- FAIL: TestPeerMultiAddrFromServiceURL_error (0.02s)\n"}
2023-02-23T16:22:46.7368388Z {"Time":"2023-02-23T16:22:46.735696151Z","Action":"output","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip6","Output":"    --- FAIL: TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip6 (0.01s)\n"}
2023-02-23T16:22:46.7369461Z {"Time":"2023-02-23T16:22:46.735699951Z","Action":"fail","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip6","Elapsed":0.01}
2023-02-23T16:22:46.7370614Z {"Time":"2023-02-23T16:22:46.735706951Z","Action":"output","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/fqdn","Output":"    --- FAIL: TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/fqdn (0.01s)\n"}
2023-02-23T16:22:46.7371673Z {"Time":"2023-02-23T16:22:46.735710751Z","Action":"fail","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/fqdn","Elapsed":0.01}
2023-02-23T16:22:46.7372953Z {"Time":"2023-02-23T16:22:46.735714351Z","Action":"output","Package":"github.com/pokt-network/pocket/libp2p/network","Test":"TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip4","Output":"    --- FAIL: TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip4 (0.01s)\n"}
```

```
2023-02-23T16:26:43.3779483Z   "Time": "2023-02-23T16:22:46.735690851Z",
2023-02-23T16:26:43.3779733Z   "Action": "output",
2023-02-23T16:26:43.3780129Z   "Package": "github.com/pokt-network/pocket/libp2p/network",
2023-02-23T16:26:43.3780511Z   "Test": "TestPeerMultiAddrFromServiceURL_error",
2023-02-23T16:26:43.3888783Z   "Output": "--- FAIL: TestPeerMultiAddrFromServiceURL_error (0.02s)\n"
2023-02-23T16:26:43.3890125Z }
2023-02-23T16:26:43.3902969Z {
2023-02-23T16:26:43.3903644Z   "Time": "2023-02-23T16:22:46.735696151Z",
2023-02-23T16:26:43.3904012Z   "Action": "output",
2023-02-23T16:26:43.3904520Z   "Package": "github.com/pokt-network/pocket/libp2p/network",
2023-02-23T16:26:43.3905066Z   "Test": "TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip6",
2023-02-23T16:26:43.3905759Z   "Output": "    --- FAIL: TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip6 (0.01s)\n"
2023-02-23T16:26:43.3906159Z }
2023-02-23T16:26:43.3906407Z {
2023-02-23T16:26:43.3906785Z   "Time": "2023-02-23T16:22:46.735699951Z",
2023-02-23T16:26:43.3907124Z   "Action": "fail",
2023-02-23T16:26:43.3907612Z   "Package": "github.com/pokt-network/pocket/libp2p/network",
2023-02-23T16:26:43.3908126Z   "Test": "TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip6",
2023-02-23T16:26:43.3908523Z   "Elapsed": 0.01
2023-02-23T16:26:43.3908800Z }
2023-02-23T16:26:43.3909032Z {
2023-02-23T16:26:43.3909413Z   "Time": "2023-02-23T16:22:46.735706951Z",
2023-02-23T16:26:43.3909762Z   "Action": "output",
2023-02-23T16:26:43.3910242Z   "Package": "github.com/pokt-network/pocket/libp2p/network",
2023-02-23T16:26:43.3910782Z   "Test": "TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/fqdn",
2023-02-23T16:26:43.3912979Z   "Output": "    --- FAIL: TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/fqdn (0.01s)\n"
2023-02-23T16:26:43.3913304Z }
2023-02-23T16:26:43.3913481Z {
2023-02-23T16:26:43.3913770Z   "Time": "2023-02-23T16:22:46.735710751Z",
2023-02-23T16:26:43.3914021Z   "Action": "fail",
2023-02-23T16:26:43.3914425Z   "Package": "github.com/pokt-network/pocket/libp2p/network",
2023-02-23T16:26:43.3914855Z   "Test": "TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/fqdn",
2023-02-23T16:26:43.3915175Z   "Elapsed": 0.01
2023-02-23T16:26:43.3915387Z }
2023-02-23T16:26:43.3915559Z {
2023-02-23T16:26:43.3916184Z   "Time": "2023-02-23T16:22:46.735714351Z",
2023-02-23T16:26:43.3916441Z   "Action": "output",
2023-02-23T16:26:43.3916895Z   "Package": "github.com/pokt-network/pocket/libp2p/network",
2023-02-23T16:26:43.3917329Z   "Test": "TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip4",
2023-02-23T16:26:43.3917915Z   "Output": "    --- FAIL: TestPeerMultiAddrFromServiceURL_error/FQDN_with_scheme/ip4 (0.01s)\n"
2023-02-23T16:26:43.3918240Z }
```

bryanchriswhite commented on shared/crypto/libp2p.go:
```diff
@@ -0,0 +1,23 @@
+func NewLibP2PPrivateKey(privateKeyHex string) (crypto.PrivKey, error) {
+	privateKeyBz, err := hex.DecodeString(privateKeyHex)
+	if err != nil {
+		return nil, ErrCreatePrivateKey(err)
```
Good idea, thanks!

bryanchriswhite commented on .github/workflows/main.yml:
```diff
@@ -55,7 +55,7 @@ jobs:
         run: |
           set -euo pipefail
-          go test -p 1 -json ./... -covermode=count -coverprofile=coverage.out 2>&1 | tee test_results.json
+          GODEBUG=netdns=cgo go test -p 1 -json ./... -covermode=count -coverprofile=coverage.out 2>&1 | tee test_results.json
```
This didn't seem to have any effect on the new test failures:

```
=== RUN   TestPeerMultiAddrFromServiceURL_Success
=== RUN   TestPeerMultiAddrFromServiceURL_Success/IPv6
    url_conversion_test.go:38: 
        	Error Trace:	/home/runner/work/pocket/pocket/libp2p/network/url_conversion_test.go:38
        	Error:      	Received unexpected error:
        	            	resolving peer IP for hostname: 2a00:1450:4005:802::2004: address 2a00:1450:4005:802::2004: no suitable address found
        	Test:       	TestPeerMultiAddrFromServiceURL_Success/IPv6
--- FAIL: TestPeerMultiAddrFromServiceURL_Success (0.00s)
    --- FAIL: TestPeerMultiAddrFromServiceURL_Success/IPv6 (0.00s)
=== RUN   TestPeerMultiAddrFromServiceURL_Error
=== RUN   TestPeerMultiAddrFromServiceURL_Error/fully_qualified_domain_name_with_scheme/ip6
    url_conversion_test.go:100: 
        	Error Trace:	/home/runner/work/pocket/pocket/libp2p/network/url_conversion_test.go:100
        	Error:      	Error "resolving peer IP for hostname: tcp: lookup tcp: Temporary failure in name resolution" does not contain "no such host"
        	Test:       	TestPeerMultiAddrFromServiceURL_Error/fully_qualified_domain_name_with_scheme/ip6
        	Messages:   	actualMultiaddr: %!s(<nil>)
=== RUN   TestPeerMultiAddrFromServiceURL_Error/fully_qualified_domain_name_with_scheme/fqdn
    url_conversion_test.go:100: 
        	Error Trace:	/home/runner/work/pocket/pocket/libp2p/network/url_conversion_test.go:100
        	Error:      	Error "resolving peer IP for hostname: tcp: lookup tcp: Temporary failure in name resolution" does not contain "no such host"
        	Test:       	TestPeerMultiAddrFromServiceURL_Error/fully_qualified_domain_name_with_scheme/fqdn
        	Messages:   	actualMultiaddr: %!s(<nil>)
=== RUN   TestPeerMultiAddrFromServiceURL_Error/fully_qualified_domain_name_with_scheme/ip4
    url_conversion_test.go:100: 
        	Error Trace:	/home/runner/work/pocket/pocket/libp2p/network/url_conversion_test.go:100
        	Error:      	Error "resolving peer IP for hostname: tcp: lookup tcp: Temporary failure in name resolution" does not contain "no such host"
        	Test:       	TestPeerMultiAddrFromServiceURL_Error/fully_qualified_domain_name_with_scheme/ip4
        	Messages:   	actualMultiaddr: %!s(<nil>)
=== RUN   TestPeerMultiAddrFromServiceURL_Error/missing_port_number/ip6
    url_conversion_test.go:100: 
        	Error Trace:	/home/runner/work/pocket/pocket/libp2p/network/url_conversion_test.go:100
        	Error:      	Error "resolving peer IP for hostname: 2a00:1450:4005:802::2004: address 2a00:1450:4005:802::2004: no suitable address found" does not contain "unexpected end of multiaddr"
        	Test:       	TestPeerMultiAddrFromServiceURL_Error/missing_port_number/ip6
        	Messages:   	actualMultiaddr: %!s(<nil>)
--- FAIL: TestPeerMultiAddrFromServiceURL_Error (0.02s)
    --- FAIL: TestPeerMultiAddrFromServiceURL_Error/fully_qualified_domain_name_with_scheme/ip6 (0.01s)
    --- FAIL: TestPeerMultiAddrFromServiceURL_Error/fully_qualified_domain_name_with_scheme/fqdn (0.01s)
    --- FAIL: TestPeerMultiAddrFromServiceURL_Error/fully_qualified_domain_name_with_scheme/ip4 (0.01s)
    --- FAIL: TestPeerMultiAddrFromServiceURL_Error/missing_port_number/ip6 (0.00s)
```

okdas replied:
I think there are a couple of things happening. First one is ipv6 being a special case, and a second one might be related to different errors on different platforms. This is just a hypothesis at this point and I need to look deeper. Do you run linux on your dev machine?

I seem to be only able to replicate ipv6 error locally and I am on mac, but I should mention my did not set up ipv6 at home.
```
❯ go test -count=1 ./libp2p/...
--- FAIL: TestPeerMultiAddrFromServiceURL_Success (0.00s)
    --- FAIL: TestPeerMultiAddrFromServiceURL_Success/IPv6 (0.00s)
        url_conversion_test.go:38:
            	Error Trace:	/Users/dk/pocket/pocket/libp2p/network/url_conversion_test.go:38
            	Error:      	Received unexpected error:
            	            	resolving peer IP for hostname: 2a00:1450:4005:802::2004: address 2a00:1450:4005:802::2004: no suitable address found
            	Test:       	TestPeerMultiAddrFromServiceURL_Success/IPv6
--- FAIL: TestPeerMultiAddrFromServiceURL_Error (15.02s)
    --- FAIL: TestPeerMultiAddrFromServiceURL_Error/missing_port_number/ip6 (0.00s)
        url_conversion_test.go:100:
            	Error Trace:	/Users/dk/pocket/pocket/libp2p/network/url_conversion_test.go:100
            	Error:      	Error "resolving peer IP for hostname: 2a00:1450:4005:802::2004: address 2a00:1450:4005:802::2004: no suitable address found" does not contain "unexpected end of multiaddr"
            	Test:       	TestPeerMultiAddrFromServiceURL_Error/missing_port_number/ip6
            	Messages:   	actualMultiaddr: %!s(<nil>)
FAIL
FAIL	github.com/pokt-network/pocket/libp2p/network	15.207s
?   	github.com/pokt-network/pocket/libp2p/transport	[no test files]
?   	github.com/pokt-network/pocket/libp2p/types/mocks	[no test files]
FAIL
```

Now, the other errors - if you read that - "resolving peer IP for hostname: tcp: lookup tcp: Temporary failure in name resolution" you can see it tries to resolve hostname `tcp` but Golang/stdlib returns `Temporary failure in name resolution` instead of `no such host`.. I suspect this happens due to inability to lookup a domain `tcp` (special case?) or the error is just different on Linux platform. I'm happy to validate/test this on fresh Linux box on Monday.

Would it be a good idea to detect ipv6/ipv6 addresses and skip DNS resolve for them? And only perform DNS request on hostnames.

Olshansk replied:
I can confirm that I'm seeing the same thing as @okdas when running this locally.

<img width="1494" alt="Screenshot 2023-02-24 at 4 07 14 PM" src="https://user-images.githubusercontent.com/1892194/221323621-2cff55d7-8c51-4aab-aa5e-a5f5fee92b49.png">

---

@bryanchriswhite does this work on your local machine?

bryanchriswhite replied:
Yes, those tests have been consistently passing locally on my Linux workstation.

bryanchriswhite replied:
EDIT: there were two tests that I apparently broke at some point and didn't notice.

Olshansk reviewed (changes requested):
With the exception of the ipv6 error discussed [here](https://github.com/pokt-network/pocket/pull/534/files#r1117231526), this change LGTM.

@bryanchriswhite Does this work for you locally? I think @okdas is probably the best person to help resolve this.

bryanchriswhite commented on libp2p/network/url_conversion.go:
```diff
@@ -0,0 +1,133 @@
+	return fmt.Errorf(errResolvePeerIPMsg, hostname, err)
+}
+
+func getPeerIP(hostname string) (net.IP, error) {
```
@okdas, @Olshansk thanks for taking a look, for adding more context, and the advice! :pray: :raised_hands: 

> Would it be a good idea to detect ipv6/ipv6 addresses and skip DNS resolve for them? And only perform DNS request on hostnames.

Yes, absolutely! I've factored out the IP resolution portion to this helper, `getPeerIP` which does this. Additionally, it's now using [`net.LookupHost`](https://pkg.go.dev/net@go1.19.5#LookupHost) (instead of [`net.ResolveIPAddr`](https://pkg.go.dev/net@go1.19.5#ResolveIPAddr)) for DNS resolution which puts us the driver seat with respect to what we want to do with a DNS response containing multiple addresses. I added a few `CONSIDER`s about this (lines 121-123).

Olshansk replied:
> which puts us the driver seat with respect to DNS responses

Wdym by this?

bryanchriswhite replied:
I meant that we get a list of all addresses in the response and get to decide what to do with it, as opposed to being at the whim of [`net.ResolveIPAddr`](https://pkg.go.dev/net#ResolveIPAddr).

> The address parameter can use a host name, but this is not recommended, because it will return at most one of the host name's IP addresses.

bryanchriswhite replied:
EDIT: if fixed my typo, "... puts us **in** the driver seat ...". Sorry for the confusion.

Olshansk replied:
Makes more sense. Ty! 

Any chance you can add a small comment on why we're using `net.ResolveIPAddr` instead of `netlLookupHost` directly in the code?

okdas replied:
Though having multiple A DNS records for the same hostname should not happen often in our case, I can see a scenario when nodes are provisioned behind a multi-zone LB, which is often a default for Kubernetes set up in clouds (I suppose they make more 🤑 that way). 

Do you think there might be issues when the same node has a hostname that resolves to multiple IP addresses?

I assume PeerID has no correlation with Multiaddr. I suppose it should be fine if one PeerID has different addresses. Imagine a situation with multiple network interfaces, e.g. internet and local networks on one machine, both interfaces have different addresses in that case. But the same node can be exposed on both interfaces. In that case, (I assume) the PeerID stays the same, but other peers should be able to connect to it just fine even if IP addresses are different. Not sure if that makes sense, I'm just kind of fascinated by this topic and my guess is random address should work as long as packets flow correctly. :)

okdas replied:
> puts us in the driver seat

Do you mean something like we could sort the IP addresses and always use the first one so all nodes in the network predictably connect to the same IP even if there are multiple A DNS records?

bryanchriswhite replied:
@okdas 

> Do you think there might be issues when the same node has a hostname that resolves to multiple IP addresses?

I think it depends on some things, I think [the canvas](https://github.com/pokt-network/pocket/pull/534#discussion_r1119808016) I put together tries to get at this.

> I assume PeerID has no correlation with Multiaddr. 

Correct, mapping from libp2p to pocket nomenclature: `PeerID` is analagous to `Address`, and `Multiaddr` to `ServiceURL`.

> I suppose it should be fine if one PeerID has different addresses.

Yes, I believe this is the case. Libp2p's [`peerstore.Peerstore`](https://pkg.go.dev/github.com/libp2p/go-libp2p@v0.26.1/core/peerstore#Peerstore) interface embeds [`peerstore.AddrBook`](https://pkg.go.dev/github.com/libp2p/go-libp2p@v0.26.1/core/peerstore#AddrBook) which makes it clear via its API that multiple multiaddrs can be associated with a given `PeerID`.

> Not sure if that makes sense, I'm just kind of fascinated by this topic and my guess is random address should work as long as packets flow correctly. :)

Indeed, that does make sense, and I agree! You've also actually called out something that I didn't completely think through, which is randomizing the selected DNS record. I was just selecting the first valid record which I'm realizing would not serve the high availability goal. :facepalm: :sweat_smile: 53067bd should address this.

> Do you mean something like we could sort the IP addresses and always use the first one so all nodes in the network predictably connect to the same IP even if there are multiple A DNS records?

Given the above, I would question the value in each node having the same topological view of the overlay network.

I just meant that we have control as opposed to being limited by [`net.ResolveIPAddr`](https://pkg.go.dev/net#ResolveIPAddr), which was the case before. It would only return a single record given a hostname:

> ... The address parameter can use a host name, but this is not recommended, because it will return at most one of the host name's IP addresses.

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -0,0 +1,133 @@
+
+	// Check IP version.
+	if peerIP.To4() == nil {
+		peerIPVersionStr = "ip6"
```
```suggestion
		peerIPVersionStr = ipVersion6
```

bryanchriswhite replied:
Nice catch, thank you! :pray:

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -0,0 +1,133 @@
+	/* CONSIDER: using a `/dns<4 or 6>/<hostname>` multiaddr instead of resolving here.
+	 * I attempted using `/dns4/.../tcp/...` and go this error:
+	 * > failed to listen on any addresses: [can only dial TCP over IPv4 or IPv6]
+	 *
```
```suggestion
```

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -0,0 +1,133 @@
+	// are provided in a DNS response?
+	// CONSIDER: preferring IPv6 responses when resolving DNS.
+	// Return first address which is a parsable IP address.
+	for _, addr := range addrs {
```
Thoughts on potentially aggregating them and logging a warning message saying "We got N addresses, returning idx 0" so its obvious if we ever get into a "wth is going on" mode?

bryanchriswhite replied:
I think it's reasonable to temporarily add a warning iff in the case where we have multiple IPs in the response. I would expect that we remove this log line though once we've implemented a longer-term solution (or made the decision not to do so). 

This begs the question: "Do we want to support multiple addresses?" However, I think it's probably more useful to come at the question from the direction of the use case. We of course, also need to think through the consequences.

Base on my current understanding, it seems like it would be appropriate to categorize running multiple nodes with the same identity as an operator error as the current implementation only holds a single service URL per peer address.

I threw together a canvas in obsidian to organize my thinking a bit:
![mult-ip](https://user-images.githubusercontent.com/600733/221813853-004ef1d9-e2d2-4481-9703-ef406f993828.png)

Olshansk replied:
This is great thinking. What you're describing is pretty much what the poktscan team implemented with GeoMesh in v0. More details here: https://forum.pokt.network/t/poktscan-s-geo-mesh/3585

We're not going to solve it now, but what I'd ask is:
1. Create a new ticket (for mileston M5) to track/thinking about this.
2. Include your diagram and a link to Geomesh, and a reference to this chat
3. If you haven't already, read through the idea of `GeoZone` staking from the utility spec - https://github.com/pokt-network/pocket-network-protocol/tree/main/utility#313-geozone
4. Let's keep the thoughts going as more ideas come up. At the end of the day, our goal is to ** Incentivize geo distributed staking with a high QoS **

bryanchriswhite replied:
https://github.com/pokt-network/pocket/issues/557

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -127,7 +127,30 @@ func getPeerIP(hostname string) (net.IP, error) {
 		if peerIP == nil {
 			continue
 		}
+		// TECHDEBT: remove this log line once direction is clearer
```
Insert link to ticket `TECHDEBT(#XXX): ...` once created

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -127,7 +127,30 @@ func getPeerIP(hostname string) (net.IP, error) {
+	strs []string
+}
+
+func (marshaler stringLogArrayMarshaler) MarshalZerologArray(arr *zerolog.Array) {
```
Where is this being used?

bryanchriswhite replied:
This is called by `zerolog` and is required by the `zerolog.LogArrayMarshaler` interface, which is the type of the second argument to [`zerolog.Evt#Array`](https://pkg.go.dev/github.com/rs/zerolog@v1.27.0#section-readme). 

I don't have a strong opinions on doing it this way, this just seemed to me like the most appropriate way to convey the information available via our structured logging. I've simplified as you suggested and added comments so nobody else has to ask the same question. :sweat_smile:

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -127,7 +127,30 @@ func getPeerIP(hostname string) (net.IP, error) {
 	return nil, newResolvePeerIPErr(hostname, err)
 }
+
+type stringLogArrayMarshaler struct {
```
Since we created this struct for a super small log line, please define it inline.

Olshansk commented on libp2p/network/url_conversion.go:
```diff
@@ -127,7 +127,30 @@ func getPeerIP(hostname string) (net.IP, error) {
+		if len(addrs) > 1 {
+			logger.Global.Warn().
+				Array("resolved", newStringLogArrayMarshaler(addrs)).
+				IPAddr("using", peerIP)
```
Can you add a message saying "resolved multiple addresses but only using one. See ticket #XXX for more details"

Olshansk commented:
@bryanchriswhite Going to take a look at this tomorrow, but I think it should be good to go after the next set of changes.

okdas reviewed (commented):
Thank you for leaving thorough comments in your code - that definitely helps me to understand what's going on! :)

bryanchriswhite commented:
@Olshansk, @okdas thanks again for taking the time! :pray:

Apologies for changing things at the last minute, but as was surfaced [in discussion with @okdas](https://github.com/pokt-network/pocket/pull/534/#discussion_r1121395427), we should be selecting a random record from the DNS response until we decide to support multiple IPs natively. As a result, There's a bit more code in `getPeerIP` than there was previously which collects the valid IPs before selecting one randomly.

I've added a test for the error case of `getPeerIP` where the resolution contained no records. I've also added a skipped test for the success cases of `getPeerIP`, skipped because to implement would require more time with [`net.Resolver`](https://pkg.go.dev/net#Resolver) and understanding how to get gomock to generate mocks for stdlib interfaces (e.g. [`net.Conn`](https://pkg.go.dev/net#Conn)).

okdas reviewed (approved):
Thank you!

I am not 100% sure if you can merge the PR unless @Olshansk reviews again due to his prior review, maybe resolving all conversation can help? Otherwise, I should be able to bypass.

Olshansk commented on libp2p/network/url_conversion_test.go:
```diff
@@ -163,9 +164,51 @@ func TestServiceUrlFromLibp2pMultiaddr_Error(t *testing.T) {
+func TestGetPeerIP_Success(t *testing.T) {
+	t.Skip("TODO: replace `net.DefaultResolver` with one which has a `Dial` function that returns a mocked `net.Conn` (see: https://pkg.go.dev/net#Resolver)")
+
+	//nolint:gocritic // commentedOutCode - Outlines the minimum requirements for disproving regression.
```
👍
